### PR TITLE
#5 Add Freebusy parsing

### LINKED ISSUE #544: [TASK] Add Freebusy parsing
Data structure should map VFREEBUSY info

#### Prerequisites

None

#### Subtasks

- [ ] Add a Freebusy data structure
- [ ] Add freebusy parsing

### PR DESCRIPTION
**Pull request by rezk2ll**
closes https://github.com/OpenPaaS-Suite/esn-frontend-calendar/issues/544

### CODE REVIEW

File: src/lib/parser.ts [hunk 1]
```diff
@@ -118,3 +118,66 @@ const parseJcalAlarm = (component: Component): Dictionary => {
 
   return alarmValue;
 }
+
+/**
+ * Parse a freebusy reply
+ *
+ * @param ics string, the vcalendar string containing the vfreebusy component
```
Comment (itswillta) [inline src/lib/parser.ts]: Isn't it supposed to be like this?

```suggestion
 * @param {string} ics - The vcalendar string containing the vfreebusy component
```

File: src/lib/parser.ts [hunk 2]
```diff
@@ -118,3 +118,66 @@ const parseJcalAlarm = (component: Component): Dictionary => {
...
+ * @returns FreeBusy[]
```
Comment (itswillta) [inline src/lib/parser.ts]:
```suggestion
 * @returns {FreeBusy[]}
```

And this?

Comment (itswillta) [reply src/lib/parser.ts]: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

File: src/lib/parser.ts (hunk 1 above)
Comment (itswillta) [reply src/lib/parser.ts]: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html

Comment (rezk2ll) [reply src/lib/parser.ts]: done, i'll update the old functions later on in a separate PR

File: src/lib/parser.ts [hunk 3]
```diff
@@ -118,3 +118,66 @@ const parseJcalAlarm = (component: Component): Dictionary => {
 
   return alarmValue;
 }
+
+/**
+ * Parse a freebusy reply
+ *
+ * @param {String} ics - the vcalendar string containing the vfreebusy component
```
Comment (itswillta) [inline src/lib/parser.ts]:
```suggestion
 * @param {string} ics - the vcalendar string containing the vfreebusy component
```

File: src/lib/parser.ts [hunk 4]
```diff
@@ -118,3 +118,66 @@ const parseJcalAlarm = (component: Component): Dictionary => {
...
+ * @returns {FreeBusy[]}
+ */
+export function parseFreeBusy(ics: string): FreeBusy[] {
+  const jcal = parseIcs(ics);
+  const vcalendarComponent = new Component(jcal);
+  const freeBusyComponents = vcalendarComponent.getAllSubcomponents('vfreebusy');
+
+  if (!freeBusyComponents) throw new Error('no busy time information were found');
```
Comment (itswillta) [inline src/lib/parser.ts:133]:
```suggestion
  if (!freeBusyComponents) throw new Error('No busy time information was found');
```

File: src/lib/parser.ts [hunk 5]
```diff
@@ -118,3 +118,66 @@ const parseJcalAlarm = (component: Component): Dictionary => {
...
+ * @param {ICAL.Component} component
+ * @returns {FreeBusy}
+ */
+const parseFreeBusyComponent = (component: Component): FreeBusy => {
+  const props: Property[] = component.getAllProperties();
+  const freeBusyObject: Dictionary = {};
+
+  props.forEach(prop => {
+    switch (prop.name) {
+    case 'attendee':
```
Comment (itswillta) [inline src/lib/parser.ts:149]: Wrong indentation.

Comment (rezk2ll) [reply src/lib/parser.ts:149]: that conflicts with the `eslint` rules setup in the library
image

**Review by renaudboyer — APPROVED** [review]

**Pull request merged**